### ISSUE TITLE
Investigate use of c_ptrTo in SPLATT

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#9883]
[SPLATT](https://github.com/thomasrolinger/Chapel-SPLATT) drops down to uses C arrays for some performance improvements. Slide 19 of https://chapel-lang.org/CHIUW/2018/rollinger-slides.pdf (and the paper) have more information.

Something like:

```chpl
var mat [0..rows-1, 0..cols-1] = ...;
for j in 0..cols-1 {
  row[i,j]...
}
```

was changed to:

```chpl
var mat [0..rows-1, 0..cols-1] = ...;
var matPtr = c_ptrTo(mat);
var row = matPtr + (i*cols);
for j in 0..cols-1 {
  row[j];
}
```

for a "1.26x speed up over 2D indexing".  We should dig into this and see what we can do to optimize the Chapel case so that dropping to a lower-level isn't necessary. At the very least it might be good to provide a dense 0-based Chapel array that would be a thin wrapper over C arrays (or more likely out _ddata) so that you wouldn't have to drop down to a c_ptr.

**Comment by cassella** [comment]
Is the "before" code meant to use `mat[i,j]`?

**Comment by cassella** [comment]
Does the `c_array` type satisfy that last sentence?

**Comment by ronawho** [comment]
I don't think so. c_array is stack allocated only and I'm not sure if you can have a c_array of wide things. 

I think this was a case where meta-data wasn't being hoisted or something, and we really just want the compiler (or llvm backend) to get a lot better at hoisting our array meta-data.

Hmm, unfortunately it looks like the SPLATT repo was removed. @thomasrolinger does the splatt code still exists anywhere?

**Comment by thomasrolinger** [comment]
@ronawho sorry about that, I must have switched the repo to private when I was cleaning things up. Should be back up now

**Comment by ronawho** [comment]
Thanks!
